src/offset/library.cairo

```diff
@@ -629,6 +479,149 @@ namespace CarbonableOffseter {
+            // Create new user
+            CarbonableOffseter_users_len_.write(index + 1);
+            CarbonableOffseter_users_.write(index, caller);
+            CarbonableOffseter_users_index_.write(caller, index);
+
+            // [Security] End reetrancy guard
+            ReentrancyGuard.end();
+            return (success=TRUE);
+        }
+
+        // [Security] End reetrancy guard
+        ReentrancyGuard.end();
+        return (success=TRUE);
+    }
+
+    func _withdraw{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        to_token_id: Uint256, to: felt, value: Uint256
+    ) -> (success: felt) {
+        alloc_locals;
+
+        // [Security] Start reetrancy guard
+        ReentrancyGuard.start();
+
+        // [Check] value is less than or equal to the registered value
+        let (caller) = get_caller_address();
+        let (registered_value) = CarbonableOffseter_registered_value_.read(caller);
+        let (is_lower) = uint256_le(value, registered_value);
+        with_attr error_message("CarbonableOffseter: value is higher than the withdrawable value") {
+            assert is_lower = TRUE;
+        }
+
+        // [Effect] Store cumulated claimable
+        let (computed_claimable) = _claimable(address=caller);
+        let (stored_claimable) = CarbonableOffseter_claimable_.read(caller);
+        CarbonableOffseter_claimable_.write(caller, stored_claimable + computed_claimable);
+
+        // [Effect] Register the caller with the new value and the current timestamp
+        let (current_time) = get_block_timestamp();
+        let (previous_value) = CarbonableOffseter_registered_value_.read(caller);
+        let (new_value) = SafeUint256.sub_le(previous_value, value);
+        CarbonableOffseter_registered_time_.write(caller, current_time);
+        CarbonableOffseter_registered_value_.write(caller, new_value);
+
+        // [Interaction] Transfer value from contract to caller
+        // let (old_deposited) = total_deposited();
+        let (from_token_id) = _get_token_id();
+        let (carbonable_project_address) = CarbonableOffseter_carbonable_project_address_.read();
+        IERC3525.transferValueFrom(
+            contract_address=carbonable_project_address,
+            fromTokenId=from_token_id,
+            toTokenId=to_token_id,
+            to=to,
+            value=value,
+        );
+
+        // [Check] Transfer successful
+        // let (new_deposited) = total_deposited();
+        // let (expected_deposited) = SafeUint256.sub_le(old_deposited, value);
+        // let (is_equal) = uint256_eq(new_deposited, expected_deposited);
+        // with_attr error_message("CarbonableOffseter: transfer failed") {
+        //     assert is_equal = TRUE;
+        // }
+
+        // [Effect] Emit event
+        Withdraw.emit(address=caller, value=value, time=current_time);
+
+        // [Effect] Remove the caller if registered value reached zero
```

Comment:
why?
Is it for accounting and efficiency?

Reply:
Yes, we will need to enumerate users while snapshoting in yielder